Objects/genericaliasobject.c

```diff
@@ -254,15 +285,46 @@ ga_getitem(PyObject *self, PyObject *item)
         if (typevar) {
             Py_ssize_t iparam = tuple_index(alias->parameters, nparams, arg);
             assert(iparam >= 0);
-            if (is_tuple) {
-                arg = PyTuple_GET_ITEM(item, iparam);
+            arg = args[iparam];
+            Py_INCREF(arg);
+        }
+        else {
```

Comment:
Can you turn this entire `else` block into a helper function? (Assuming there aren't too many variables to pass in.) Or at least a large portion of it. The top-level logic is hard to follow (I almost needed to resort to pen and paper to convince myself it is correct :-).

Also we could use a few examples, e.g.
```
t = list[T];          t[int]      -> newargs = [int]
t = dict[str, T];     t[int]      -> newargs = [str, int]
t = dict[T, list[S]]; t[str, int] -> newargs = [str, list[int]]
```